akka-http2-support/src/test/scala/akka/http/impl/engine/http2/WindowTracking.scala

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.http.impl.engine.http2
+
+import akka.http.impl.engine.http2.Http2Protocol.FrameType
+import akka.stream.impl.io.ByteStringParser.ByteReader
+import akka.util.ByteString
+
+import scala.concurrent.duration.FiniteDuration
+
+trait WindowTracking extends Http2FrameProbeDelegator with Http2FrameSending {
+  override def sendDATA(streamId: Int, endStream: Boolean, data: ByteString): Unit = {
+    updateToServerWindowForConnection(_ - data.length)
+    updateToServerWindows(streamId, _ - data.length)
+    super.sendDATA(streamId, endStream, data)
+  }
+
+  override def sendWINDOW_UPDATE(streamId: Int, windowSizeIncrement: Int): Unit = {
+    super.sendWINDOW_UPDATE(streamId, windowSizeIncrement)
+    if (streamId == 0) updateFromServerWindowForConnection(_ + windowSizeIncrement)
+    else updateFromServerWindows(streamId, _ + windowSizeIncrement)
+  }
+
+  def expectWindowUpdate(): Unit =
+    expectFrameFlagsStreamIdAndPayload(FrameType.WINDOW_UPDATE) match {
+      case (flags, streamId, payload) =>
+        // TODO: DRY up with autoFrameHandler
+        val windowSizeIncrement = new ByteReader(payload).readIntBE()
+
+        if (streamId == 0) updateToServerWindowForConnection(_ + windowSizeIncrement)
+        else updateToServerWindows(streamId, _ + windowSizeIncrement)
+    }
+
+  final def pollForWindowUpdates(duration: FiniteDuration): Unit =
+    try {
+      within(duration)(expectWindowUpdate())
+
+      pollForWindowUpdates(duration)
+    } catch {
+      case e: AssertionError if e.getMessage contains "but only got [0] bytes" =>
+      // timeout, that's expected
+      case e: AssertionError if (e.getMessage contains "block took") && (e.getMessage contains "exceeding") =>
+        // pause like GC, poll again just to be sure
+        pollForWindowUpdates(duration)
+    }
+
+  // keep counters that are updated on outgoing network.sendDATA and incoming WINDOW_UPDATE frames
+  private var toServerWindows: Map[Int, Int] = Map.empty.withDefaultValue(Http2Protocol.InitialWindowSize)
```

Comment:
These names are now not valid anymore. Maybe rename all those with "server" in it to `toPeerWindow...`

Reply:
Naming is hard! :D

This represents the window we're publishing for incoming data, so perhaps `windowForIncomingData`.